File: haskell/src/Problem_004.hs
-- Find the largest palindrome made from the product of two 3-digit numbers.

module Main where
import Data.List

isPalindrome xs = xs == (reverse xs)
main = print (last (sort [x * y| x <- [900..999], y <- [x..999], isPalindrome (show (x * y))]))

File: scala/src/Problem_004.scala

/** Find the largest palindrome made from the product of two 3-digit numbers. */
object Problem_004 extends App {
  import math.max

  def isPalindrome(x: Int) = x.toString match {
    case s: String => s == s.reverse
  }

  def result = (for (x <- 900 to 999; y <- x to 999; c = x * y if isPalindrome(c)) yield c).foldLeft(0)(max)

  assert(906609 == result)
}